scalable-go-service/index.md

```diff
@@ -0,0 +1,322 @@
+was not immediately obvious how these components would run independently. The
+rate loader takes a validated request from the route handler and needs a token
+before contacting the vendor. However, we noticed that multiple rate loaders can
+execute independently once they have all of the necessary info. Once we
+facilitate proper communication between the components, we can start regarding
+them as workers that receive required parameters from `channels`, perform their
+task, and communicate the result through `channels`. Groups of workers that
+perform the same tasks can be regarded as a pool of workers. They all listen on
+a single channel for requests they can process.
+
+<p align="center">
+    <img src="./pictures/third.jpg" width="600" heigth:"400"/>
+</p>
+
+
+How many rate loaders? It depends on the application, but this is now a
+configurable parameter. It does not have to be an exact number either. One might
+configure a lower and an upper bound for how many rate loaders they want in
+their app and can add logic to dynamically scale to throw limits based on the
+current load.
+
+## Sample Code
+
+#### Main
+
+```go
+func main() {
+	var wg sync.WaitGroup
+	tokenRQChan := make(chan domain.TokenRQ)
+	ratesRQChan := make(chan domain.RateRQ, rateLoaderCount*2)
+
+	var tokenLoader = TokenLoader{
+		In:   tokenRQChan,
+	}
+	var rateLoaderPool = RateLoaderPool{
+		Count:         viper.GetInt("rateLoaderCount"),
+		In:            ratesRQChan,
+		TokenRQChan:   tokenRQChan,
+	}
+
+	appRouter := mux.NewRouter()
+	appRouter.HandleFunc("/rates", RatesHandler(ratesRQChan))
+	srv := http.Server{
+		Addr:    viper.GetString("server.host") + ":" + viper.GetString("server.port"),
+		Handler: appRouter,
+	}
+
+	wg.Add(1)
+	go func() {
+		tokenLoader.Start()
+		wg.Done()
+	}()
+
+	wg.Add(1)
+	go func() {
+		rateLoaderPool.Start()
+		wg.Done()
+	}()
+
+    // omitting implementation of gracefully shutting down all the workers
+}
+```
+
+#### Rates Route Handler
+
+```go
+func RatesHandler(ratesRQChan chan domain.RateRQ) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+	    var rates []types.HotelRateInfo
+	    rateRSChan := make(chan RateRS, len(r.HotelIds))
+	    for _, hotelID := range r.HotelIds {
+	    	ratesRQChan <- RateRQ{
+	    		HotelID:                 hotelID,
+	    		CheckInDate:             r.CheckInDate,
+	    		CheckOutDate:            r.CheckOutDate,
+	    		RateRSChan:              rateRSChan,
+	    	}
+	    }
+	    for range p.HotelIds {
+	    	res := <-rateRSChan
+            rates = append(rates, res.HotelRateInfo)
+	    }
+		writeResponse(&w, createJsonResponse(r.Context(), rates, http.StatusOK))
+	}
+}
+```
+
+#### Rate Loader Pool
+
+```go
+type RateLoaderPool struct {
+	Count         int
+	In            chan RateRQ
+	TokenRQChan   chan<- TokenRQ
+	stop          chan struct{}
+}
+
+// Start initializes and starts n unique rate loader based on the Count parameter.
```

Comment:
Comments are weird in a blog post. Either the code is self explanatory or the surrounding paragraphs give enough context.